src/runner.ts

```diff
@@ -1,3 +1,102 @@
+import * as dopri from "dopri";
+import type { UserType } from "./user";
+import * as userHelpers from "./user";
+
+// Probably this is something that dopri should export for us, we
+// could also use its types for the rhs and output members below.
+type Solution = (t: number) => number[];
+
+export type OdinModelConstructable =
+    new(userHelpers: any, pars: UserType, unknownAction: string) => OdinModel;
+
+interface OdinModelODE {
+    initial(t: number): number[];
+    rhs(t: number, y: number[], dydt: number[]): void;
+    output?(t: number, y: number[]): number[];
+    names(): string[];
+}
+
+interface OdinModelDDE {
+    initial(t: number): number[];
+    rhs(t: number, y: number[], dydt: number[], solution: Solution): void;
+    output?(t: number, y: number[], solution: Solution): number[];
+    names(): string[];
+}
+
+export type OdinModel = OdinModelODE | OdinModelDDE;
+
+export function delay(solution: Solution, t: number, index: number[],
+                      state: number[]) {
+    // Later, we'll update dopri.js to allow passing index here,
+    // which will make this more efficient. However, no change to
+    // the external interface will be neeed.
+    const y = solution(t);
+    for (let i = 0; i < index.length; ++i) {
+        state[i] = y[index[i]];
+    }
+}
+
+export const base = {checkUser: userHelpers.checkUser,
+                     delay,
+                     getUserScalar: userHelpers.getUserScalar};
+
+// tslint:disable-next-line:variable-name
+export function wodinRun(Model: OdinModelConstructable, pars: UserType,
+                         tStart: number, tEnd: number,
+                         control: any) {
+    const model = new Model(base, pars, "error");
+    const solution = (
+        model.rhs.length === 4 ?
```

Comment:
the more Typescript-y way of doing this would be with a custom type guard, something like:

```
function isDDEModel(model: OdinModelConstructable): model is OdinModelDDe {
   return model.rhs.length === 4
}
```

then here:
```
  const solution =
        isDDEModel(model) ?
            wodinRunDDE(model, tStart, tEnd, control) :
            wodinRunODE(model, tStart, tEnd, control);

```

Reply:
thanks, added that